src/pip/_internal/cache.py

```diff
@@ -185,6 +182,7 @@ def __init__(self, cache_dir, format_control):
     def get_path_for_link_legacy(self, link):
         # type: (Link) -> str
         parts = self._get_cache_path_parts_legacy(link)
+        assert self.cache_dir
```

Comment:
Any object a caller has to verify is OK before calling some function is just asking to be misused. At least with these changes it is explicit!